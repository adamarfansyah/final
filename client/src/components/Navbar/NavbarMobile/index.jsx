import PropTypes from 'prop-types';
import { useState } from 'react';

import MenuIcon from '@mui/icons-material/Menu';
import LightModeIcon from '@mui/icons-material/LightMode';
import NightsStayIcon from '@mui/icons-material/NightsStay';
import { Link, useLocation } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
import { Avatar, Menu, MenuItem, ToggleButton, ToggleButtonGroup } from '@mui/material';

import Button from '@components/Button';
import classes from './style.module.scss';

const NavbarMobile = ({
  links,
  navigate,
  user,
  onSelectLang,
  locale,
  handleTheme,
  theme,
  alignment,
  handleAlignment,
  handleLogout,
}) => {
  const location = useLocation();
  const [menuPosition, setMenuPosition] = useState(null);

  const open = Boolean(menuPosition);

  const handleClick = (event) => {
    setMenuPosition(event.currentTarget);
  };

  const handleClose = () => {
    setMenuPosition(null);
  };

  return (
    <div>
      <div onClick={handleClick}>
        <div className={classes.toggle}>
          <MenuIcon />
        </div>
      </div>
      <div className={classes.toolbar}>
        <Menu open={open} anchorEl={menuPosition} onClose={handleClose}>
          <MenuItem>
            <ToggleButtonGroup size="small" value={alignment} exclusive onChange={handleAlignment}>
              <ToggleButton
                value="left"
                aria-label="left aligned"
                onClick={() => onSelectLang('id')}
                selected={locale === 'id'}
              >
                <div className={classes.locale}>
                  <Avatar className={classes.avatar} src="/id.png" />
                  <div className={classes.lang}>ID</div>
                </div>
              </ToggleButton>
              <ToggleButton
                value="center"
                aria-label="centered"
                onClick={() => onSelectLang('en')}
                selected={locale === 'en'}
              >
                <div className={classes.locale}>
                  <Avatar className={classes.avatar} src="/en.png" />
                  <div className={classes.lang}>EN</div>
                </div>
              </ToggleButton>
            </ToggleButtonGroup>
          </MenuItem>
          <MenuItem onClick={handleTheme}>
            <div className={classes.theme} data-testid="toggleTheme">
              {theme === 'light' ? <NightsStayIcon /> : <LightModeIcon />}
              <div>{theme === 'light' ? 'Light' : 'Dark'}</div>
            </div>
          </MenuItem>
          {Object.keys(user).length > 0 && (
            <MenuItem onClick={() => navigate('/profile')}>
              <div className={classes.userNav}>
                <img className={classes.imgUser} src={user.image} alt={user.firstName} />
                {user.firstName}
              </div>
            </MenuItem>
          )}
          {links?.map((link, idx) => (
            <MenuItem key={idx}>
              <Link
                className={location.pathname !== link.href ? classes.link : `${classes.link} ${classes.active}`}
                key={link.id}
                to={link.href}
              >
                <FormattedMessage id={link.name} />
              </Link>
            </MenuItem>
          ))}
          {Object.keys(user).length > 0 ? (
            <MenuItem>
              <Button className={classes.btn} onClick={() => handleLogout()}>
                <FormattedMessage id="navbar_link_merchant_logout" />
              </Button>
            </MenuItem>
          ) : (
            <>
              <MenuItem>
                <Button className={classes.btn} onClick={() => navigate('/auth')}>
                  <FormattedMessage id="app_login_title" />
                </Button>
              </MenuItem>
              <MenuItem>
                <Button className={classes.btn} onClick={() => navigate('/auth/merchant')}>
                  <FormattedMessage id="app_login_title_merchant" />
                </Button>
              </MenuItem>
            </>
          )}
        </Menu>
      </div>
    </div>
  );
};

NavbarMobile.propTypes = {
  navigate: PropTypes.func,
  user: PropTypes.object,
  onSelectLang: PropTypes.func,
  locale: PropTypes.string,
  handleTheme: PropTypes.func,
  theme: PropTypes.string,
  alignment: PropTypes.string,
  handleAlignment: PropTypes.func,
  links: PropTypes.array,
  handleLogout: PropTypes.func,
};

export default NavbarMobile;
