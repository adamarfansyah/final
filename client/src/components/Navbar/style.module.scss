@import '@styles/base';

.headerWrapper {
  background-color: var(--color-bg);
  box-shadow: 0 0.25rem 0.75rem rgba(25, 49, 83, 0.2) !important;
  position: sticky;
  top: 0;
  transition: 0.3s;
  z-index: 999;
  &.headerScrolling {
    background: rgba(255, 255, 255, 0.12);
    backdrop-filter: blur(5.3px);
    -webkit-backdrop-filter: blur(5.3px);
  }
  .contentWrapper {
    @include directionX;
    color: var(--color-text);
    padding: 1rem;
    align-items: center;
    justify-content: space-between;
    font-size: 1rem;
    .logoImage {
      @include directionX;
      cursor: pointer;
      align-items: center;
      .logo {
        width: 1.5rem;
        height: 1.5rem;
        margin-right: 0.5rem;
      }
      .title {
        font-weight: 700;
      }
    }
  }

  .toolbar {
    @include directionX;
    .theme {
      @include directionX;
      align-items: center;
      margin-right: 0.5rem;
      cursor: pointer;
    }
  }
  .rightContent {
    @include directionX;
    align-items: center;
    gap: 30px;
    .toggle {
      font-size: 30px;
      cursor: pointer;
    }
    .menus {
      @include directionX;
      gap: 20px;
      .link {
        color: var(--color-text);
        text-decoration: none;
        font-weight: bold;
        font-size: 16px;
        letter-spacing: 1px;
        transition: color 0.2s;
        &:hover {
          color: var(--color-button);
        }
        &.active {
          border-bottom: 2px solid var(--color-button);
          color: var(--color-button);
        }
      }
    }
  }
}

.locale {
  @include directionX;
  align-items: center;
  .avatar {
    width: 1.2rem;
    height: 1.2rem;
    box-shadow: rgba(99, 99, 99, 0.2) 0px 2px 8px 0px;
    margin-right: 0.5rem;
  }
  .lang {
    font-family: Poppins, Arial, Helvetica, sans-serif;
    font-weight: 700;
    font-size: 0.6rem;
    text-transform: uppercase;
  }
}

.theme {
  @include directionX;
  gap: 10px;
}

// .headerWrapper {
//   background-color: var(--color-bg) !important;
//   box-shadow: 0 0.25rem 0.75rem rgba(25, 49, 83, 0.2) !important;
//   position: sticky;
//   top: 0;
//   .contentWrapper {
//     @include directionX;
//     color: var(--color-text);
//     padding: 1rem;
//     align-items: center;
//     justify-content: space-between;
//     font-size: 1rem;

//     .logoImage {
//       @include directionX;
//       cursor: pointer;
//       align-items: center;

//       .logo {
//         width: 1.5rem;
//         height: 1.5rem;
//         margin-right: 0.5rem;
//       }
//       .title {
//         font-weight: 700;
//       }
//     }

//     .toolbar {
//       @include directionX;
//       .theme {
//         @include directionX;
//         align-items: center;
//         margin-right: 0.5rem;
//         cursor: pointer;
//       }
//       .toggle {
//         @include directionX;
//         align-items: center;
//         cursor: pointer;

//         .avatar {
//           width: 1.2rem;
//           height: 1.2rem;
//           box-shadow: rgba(99, 99, 99, 0.2) 0px 2px 8px 0px;
//           margin-right: 0.5rem;
//         }

//         .lang {
//           font-family: Poppins, Arial, Helvetica, sans-serif;
//           font-weight: 700;
//           font-size: 1rem;
//           text-transform: uppercase;
//         }

//         .arrow {
//           font-size: 1rem;
//         }
//       }
//     }
//   }
// }

// .menu {
//   @include directionX;
//   align-items: center;

//   .menuAvatar {
//     width: 1rem;
//     height: 1rem;
//     box-shadow: rgba(99, 99, 99, 0.2) 0px 2px 8px 0px;
//     margin-right: 0.5rem;
//   }

//   .menuLang {
//     font-family: Poppins, Arial, Helvetica, sans-serif;
//     font-size: 0.8rem;
//   }
// }
