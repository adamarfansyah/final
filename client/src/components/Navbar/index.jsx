import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { FormattedMessage, injectIntl } from 'react-intl';
import { useDispatch } from 'react-redux';
import { Link, useLocation, useNavigate } from 'react-router-dom';
import Avatar from '@mui/material/Avatar';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import LightModeIcon from '@mui/icons-material/LightMode';
import NightsStayIcon from '@mui/icons-material/NightsStay';
import MenuIcon from '@mui/icons-material/Menu';

import { ToggleButton, ToggleButtonGroup, useMediaQuery } from '@mui/material';
import { setLocale } from '@containers/App/actions';
import { logoutUser } from '@pages/Auth/actions';
import Button from '@components/Button';
import NavbarMobile from './NavbarMobile';
import classes from './style.module.scss';

const Navbar = ({ locale, user, theme, links, handleTheme }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const location = useLocation();
  const isMobile = useMediaQuery('(max-width:600px)');
  const [menuPosition, setMenuPosition] = useState(null);
  const [alignment, setAlignment] = useState('left');
  const [scrolling, setScrolling] = useState(false);

  useEffect(() => {
    const handleScroll = () => {
      const isScrolling = window.scrollY > 100;
      setScrolling(isScrolling);
    };

    window.addEventListener('scroll', handleScroll);

    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, []);

  const open = Boolean(menuPosition);

  const handleClick = (event) => {
    setMenuPosition(event.currentTarget);
  };

  const handleClose = () => {
    setMenuPosition(null);
  };

  const handleLogout = () => {
    dispatch(logoutUser());
  };

  const onSelectLang = (lang) => {
    if (lang !== locale) {
      dispatch(setLocale(lang));
    }
    handleClose();
  };

  const goTo = (href) => {
    navigate(href);
  };

  const handleAlignment = (newAlignment) => {
    if (newAlignment !== null) {
      setAlignment(newAlignment);
    }
  };

  return (
    <div
      className={`${!scrolling ? classes.headerWrapper : `${classes.headerWrapper} ${classes.headerScrolling}`}`}
      data-testid="navbar"
    >
      <div className={classes.contentWrapper}>
        <div className={classes.logoImage} onClick={() => goTo('/')}>
          <img src="/logoPlayZone.png" alt="logo" className={classes.logo} />
          <div className={classes.title}>PlayZone Pro</div>
        </div>
        {!isMobile ? (
          <div className={classes.rightContent}>
            <div className={classes.menus}>
              {links?.map((link) => (
                <Link
                  className={location.pathname !== link.href ? classes.link : `${classes.link} ${classes.active}`}
                  key={link.id}
                  to={link.href}
                >
                  <FormattedMessage id={link.name} />
                </Link>
              ))}
              {user && user.id ? (
                <>
                  <Link
                    className={location.pathname !== '/profile' ? classes.link : `${classes.link} ${classes.active}`}
                    to="/profile"
                  >
                    <FormattedMessage id="navbar_link_merchant_profile" />
                  </Link>
                  <Button onClick={handleLogout}>
                    <FormattedMessage id="navbar_link_merchant_logout" />
                  </Button>
                </>
              ) : (
                <>
                  <Button onClick={() => goTo('/auth')}>
                    <FormattedMessage id="app_login_title" />
                  </Button>
                  <Button onClick={() => goTo('/auth/merchant')}>
                    <FormattedMessage id="app_login_title_merchant" />
                  </Button>
                </>
              )}
            </div>
            <div className={classes.toolbar}>
              <div onClick={handleClick}>
                <MenuIcon className={classes.toggle} />
              </div>
              <Menu open={open} anchorEl={menuPosition} onClose={handleClose}>
                <MenuItem>
                  <ToggleButtonGroup size="small" value={alignment} exclusive onChange={handleAlignment}>
                    <ToggleButton
                      value="left"
                      aria-label="left aligned"
                      onClick={() => onSelectLang('id')}
                      selected={locale === 'id'}
                    >
                      <div className={classes.locale}>
                        <Avatar className={classes.avatar} src="/id.png" />
                        <div className={classes.lang}>ID</div>
                      </div>
                    </ToggleButton>
                    <ToggleButton
                      value="center"
                      aria-label="centered"
                      onClick={() => onSelectLang('en')}
                      selected={locale === 'en'}
                    >
                      <div className={classes.locale}>
                        <Avatar className={classes.avatar} src="/en.png" />
                        <div className={classes.lang}>EN</div>
                      </div>
                    </ToggleButton>
                  </ToggleButtonGroup>
                </MenuItem>
                <MenuItem onClick={handleTheme}>
                  <div className={classes.theme} data-testid="toggleTheme">
                    {theme === 'light' ? <NightsStayIcon /> : <LightModeIcon />}
                    <div>{theme === 'light' ? 'Light' : 'Dark'}</div>
                  </div>
                </MenuItem>
              </Menu>
            </div>
          </div>
        ) : (
          <NavbarMobile
            handleAlignment={handleAlignment}
            alignment={alignment}
            onSelectLang={onSelectLang}
            locale={locale}
            theme={theme}
            handleTheme={handleTheme}
            user={user}
            navigate={goTo}
            links={links}
            handleLogout={handleLogout}
          />
        )}
      </div>
    </div>
  );
};

Navbar.propTypes = {
  locale: PropTypes.string.isRequired,
  user: PropTypes.object,
  theme: PropTypes.string,
  handleTheme: PropTypes.func,
  links: PropTypes.array,
};

export default injectIntl(Navbar);
