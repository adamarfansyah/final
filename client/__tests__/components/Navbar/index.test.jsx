import { render } from '@utils/testHelper';

import Navbar from '@components/Navbar';

const mockLocation = jest.fn();
// const mockDispatch = jest.fn();

// jest.mock('react-redux', () => ({
//   useDispatch: () => mockDispatch,
// }));

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
  useLocation: () => mockLocation,
}));

describe('Navbar Component', () => {
  test('Correct render', () => {
    const navbar = render(<Navbar title="Title" locale="en" theme="light" />);
    expect(navbar.getByTestId('navbar')).toBeInTheDocument();
  });

  test('Should match with snapshot', () => {
    const navbar = render(<Navbar title="Title" locale="en" theme="light" />);
    expect(navbar).toMatchSnapshot();
  });
});
