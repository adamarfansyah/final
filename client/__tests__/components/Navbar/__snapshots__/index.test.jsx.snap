// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="headerWrapper"
        data-testid="navbar"
      >
        <div
          class="contentWrapper"
        >
          <div
            class="logoImage"
          >
            <img
              alt="logo"
              class="logo"
              src="/logoPlayZone.png"
            />
            <div
              class="title"
            >
              PlayZone Pro
            </div>
          </div>
          <div
            class="rightContent"
          >
            <div
              class="menus"
            >
              <button
                class="btn"
                data-testid="btn"
                type="submit"
              >
                Masuk
              </button>
              <button
                class="btn"
                data-testid="btn"
                type="submit"
              >
                Masuk sebagai Mitra
              </button>
            </div>
            <div
              class="toolbar"
            >
              <div>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium toggle css-i4bv87-MuiSvgIcon-root"
                  data-testid="MenuIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="headerWrapper"
      data-testid="navbar"
    >
      <div
        class="contentWrapper"
      >
        <div
          class="logoImage"
        >
          <img
            alt="logo"
            class="logo"
            src="/logoPlayZone.png"
          />
          <div
            class="title"
          >
            PlayZone Pro
          </div>
        </div>
        <div
          class="rightContent"
        >
          <div
            class="menus"
          >
            <button
              class="btn"
              data-testid="btn"
              type="submit"
            >
              Masuk
            </button>
            <button
              class="btn"
              data-testid="btn"
              type="submit"
            >
              Masuk sebagai Mitra
            </button>
          </div>
          <div
            class="toolbar"
          >
            <div>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium toggle css-i4bv87-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
